File: backend/app/schemas/document.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


DocumentStatus = Literal[
    "uploaded",
    "queued",
    "processing",
    "ready",
    "failed",
]


class DocumentResponse(BaseModel):
    id: str
    user_id: str
    original_filename: str
    content_type: str
    file_extension: str
    file_size: int = Field(ge=0)
    checksum_sha256: str
    status: DocumentStatus
    language: str | None
    page_count: int | None
    word_count: int | None
    parsing_quality_score: float | None
    requires_ocr: bool
    error_message: str | None
    created_at: datetime
    updated_at: datetime
    processed_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class DocumentListResponse(BaseModel):
    items: list[DocumentResponse]
    total: int = Field(ge=0)
    limit: int = Field(ge=1)
    offset: int = Field(ge=0)


class DocumentUnitResponse(BaseModel):
    id: str
    document_id: str
    unit_index: int = Field(ge=1)
    unit_type: Literal[
        "page",
        "slide",
        "sheet",
        "section",
        "text",
    ]
    source_label: str
    content: str
    content_hash: str
    char_count: int = Field(ge=0)
    word_count: int = Field(ge=0)
    unit_metadata: dict[str, object]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DocumentUnitListResponse(BaseModel):
    document_id: str
    document_status: DocumentStatus
    items: list[DocumentUnitResponse]
    total: int = Field(ge=0)
    limit: int = Field(ge=1)
    offset: int = Field(ge=0)

File: frontend/src/types/document.ts
export type DocumentStatus =
  | "uploaded"
  | "queued"
  | "processing"
  | "ready"
  | "failed";

export type DocumentResponse = {
  id: string;
  user_id: string;
  original_filename: string;
  content_type: string;
  file_extension: string;
  file_size: number;
  checksum_sha256: string;
  status: DocumentStatus;
  language: string | null;
  page_count: number | null;
  word_count: number | null;
  parsing_quality_score: number | null;
  requires_ocr: boolean;
  error_message: string | null;
  created_at: string;
  updated_at: string;
  processed_at: string | null;
};

export type DocumentListResponse = {
  items: DocumentResponse[];
  total: number;
  limit: number;
  offset: number;
};

export type DocumentUnitType =
  | "page"
  | "slide"
  | "sheet"
  | "section"
  | "text";

export type DocumentUnitResponse = {
  id: string;
  document_id: string;
  unit_index: number;
  unit_type: DocumentUnitType;
  source_label: string;
  content: string;
  content_hash: string;
  char_count: number;
  word_count: number;
  unit_metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type DocumentUnitListResponse = {
  document_id: string;
  document_status: DocumentStatus;
  items: DocumentUnitResponse[];
  total: number;
  limit: number;
  offset: number;
};